File: Components/search_mu_uniform.m
function [output1,output2] = search_mu_uniform(varargin)
% The uniform mutation.

%------------------------------Copyright-----------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------

mode = varargin{end};
switch mode
    case 'execute'
        Parent  = varargin{1};
        Problem = varargin{2};
        Para    = varargin{3};
        Aux     = varargin{4};

        if ~isnumeric(Parent)
            Offspring = Parent.decs;
        else
            Offspring = Parent;
        end
        Prob  = Para;
        [N,D] = size(Offspring);
        
        Lower = Problem.bound(1,:);
        Upper = Problem.bound(2,:);
        ind = rand(N,D) < Prob;
        Temp = unifrnd(repmat(Lower,N,1),repmat(Upper,N,1));
        Offspring(ind) = Temp(ind);
        output1 = Offspring;
        output2 = Aux;

    case 'parameter'
        output1 = [0,0.3]; % mutation probability 
    case 'behavior'
        output1 = {'LS','small';'GS','large'}; % small probabilities perform local search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: AutoOptLib for Python/components/search_mu_uniform.py
"""Python translation of search_mu_uniform."""

from __future__ import annotations

import numpy as np


def search_mu_uniform(*args):
    mode = args[-1]
    if mode == "execute":
        parent = args[0]
        problem = args[1] if len(args) > 1 else None
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None  # noqa: F841

        if not isinstance(parent, (np.ndarray, list, tuple)):
            decs = getattr(parent, "decs", None)
            offspring = decs() if callable(decs) else (decs if decs is not None else parent)
        else:
            offspring = parent
        offspring = np.asarray(offspring, dtype=float)
        prob = float(np.asarray(para).reshape(-1)[0]) if para is not None else 0.1
        n, d = offspring.shape
        lower = np.asarray(getattr(problem, "bound")[0], dtype=float)
        upper = np.asarray(getattr(problem, "bound")[1], dtype=float)
        mask = np.random.rand(n, d) < prob
        rand_vals = lower + (upper - lower) * np.random.rand(n, d)
        offspring[mask] = rand_vals[mask]
        return offspring, aux

    if mode == "parameter":
        return [0, 0.3], None

    if mode == "behavior":
        return [["LS", "small"], ["GS", "large"]], None

    raise ValueError(f"Unsupported mode: {mode}")
